src/server/engine_shard_set.cc

```diff
@@ -158,6 +162,15 @@ bool EngineShard::DoDefrag() {
   } while (traverses_count < kMaxTraverses && cur);
 
   defrag_state_.UpdateScanState(cur.value());
+
+  // Once we're done with a db, jump to the next
+  if (defrag_state_.cursor == kCursorDoneState) {
+    defrag_state_.dbid++;
+    // Skip null dbs that still take up slots in the array
+    while (defrag_state_.dbid < slice.db_array_size() && !slice.IsDbValid(defrag_state_.dbid))
+      defrag_state_.dbid++;
+  }
```

Comment:
Yes, if the db has less elements than kMaxTraverses, then it will proceed with the next db only on the next phase. I don't see it as a big flaw given that `kMaxTraverses` is only 40 now.